File: 14-longest-common-prefix/longest-common-prefix.java
class Trie {
    TrieNode root;

    Trie() {
        this.root = new TrieNode();
    }

    void insert(String s) {
        TrieNode curr = this.root;
        for (Character c : s.toCharArray()) {
            if (curr.children[c - 'a'] == null) {
                curr.children[c - 'a'] = new TrieNode();
            }
            curr = curr.children[c - 'a'];
        }
        // return True;
        curr.isEnd = true;
    }

    String search() {
        TrieNode curr = this.root;
        StringBuilder sb = new StringBuilder();
        while (!curr.isEnd && cntChildren(curr) == 1) {
            for (int i = 0; i < curr.children.length; i++) {
                if (curr.children[i] != null) {
                    sb.append((char)('a' + i));
                    curr = curr.children[i];
                    break;
                }
            }
        }
        return sb.toString();
    }

    int cntChildren (TrieNode root) {
        int cnt = 0;
        for (TrieNode x : root.children) {
            if (x != null) {
                cnt += 1;
            }
        }
        return cnt;
    }
}

class TrieNode {
    TrieNode[] children = new TrieNode[26];
    boolean isEnd = false;
}

class Solution {
    public String longestCommonPrefix(String[] strs) {
        Trie t = new Trie();
        for (String s: strs) {
            t.insert(s);
        }
        return t.search();
    }
}

File: 14-longest-common-prefix/longest-common-prefix.py
class Solution:
    def longestCommonPrefix(self, strs: List[str]) -> str:
        if "" in strs:
            return ""
        class TrieNode:
            def __init__(self):
                self.children = [None] * 26
                self.isEnd = False

        class Trie:
            def __init__(self):
                self.root = TrieNode()

            def insert(self, word: str) -> None:
                curr = self.root
                for i in word:
                    idx = ord(i) - ord('a')
                    if not curr.children[idx]:
                        curr.children[idx] = TrieNode()
                    curr = curr.children[idx]
                curr.isEnd = True
                return curr

            def oneChild(self, root) -> bool:
                cnt = 0
                for i in root.children:
                    if i:
                        cnt += 1
                    if cnt > 1:
                        return False
                if cnt <= 0:
                    return False
                return True

            def startsWith(self):
                curr = self.root
                ans = ""
                # for i in prefix:
                while self.oneChild(curr) and not curr.isEnd:
                    # idx = ord(i) - ord('a')
                    for i in range(len(curr.children)):
                        if curr.children[i]:
                            idx = i
                            break
                    ans += chr(ord('a') + i)
                    if not curr.children[idx]:
                        return False
                    curr = curr.children[idx]
                    # else:
                        # break

                return ans

        t = Trie()
        for word in strs:
            t.insert(word)
        return t.startsWith()
